polygeist/lib/Dialect/Polygeist/Analysis/MemoryAccessAnalysis.cpp

```diff
@@ -641,6 +676,277 @@ MemoryAccess<OpTy>::classifyMemoryAccess(DataFlowSolver &solver) const {
   return MemoryAccessPattern::Unknown;
 }
 
+//===----------------------------------------------------------------------===//
+// MemoryAccessAnalysis
+//===----------------------------------------------------------------------===//
+
+/// Walks up the parents and records the ones with the specified type.
+template <typename T>
+static SetVector<Operation *> getParentsOfType(Block *block) {
+  SetVector<Operation *> res;
+  auto *current = block->getParentOp();
+  while (current) {
+    if (auto typedParent = dyn_cast<T>(current)) {
+      assert(res.count(current) == 0 && "Already inserted");
+      res.insert(current);
+    }
+    current = current->getParentOp();
+  }
```

Comment:
```suggestion
  auto *current = block->getParentOp();
  while (current = current->getParentOfType<T>()) {
    assert(res.count(current) == 0 && "Already inserted");
    res.insert(current);
  }
```

Reply:
as the warning suggested, just need to add `( )`, so `while ((current = current->getParentOfType<T>())) {`.

Reply:
Whitney suggestion would miss the first operation if that's of type `T`, I think.
```cpp
  constexpr auto getInitialParent = [](Block *b) -> T {
    Operation *op = b->getParent();
    return isa<T>(op) ? cast<T>(op) : op->getParentOfType<T>();
  };
  for (T parent = getInitialParent(block); parent; parent = parent->getParentOfType<T>())
    res.insert(parent);
```